src/Parallafka/CommitState.cs

```diff
@@ -0,0 +1,185 @@
+﻿using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Parallafka.KafkaConsumer;
+
+namespace Parallafka
+{
+    /// <summary>
+    /// Keeps track of messages that need to be committed.
+    /// </summary>
+    /// <typeparam name="TKey"></typeparam>
+    /// <typeparam name="TValue"></typeparam>
+    internal class CommitState<TKey, TValue>
+    {
+        /// <summary>
+        /// Maps partition number to a queue of the messages received from that partition.
+        /// The front of the queue is the earliest uncommitted message. Messages are removed once committed.
+        /// It's safe to commit a handled message provided all earlier (lower-offset) messages have also been marked
+        /// as handled.
+        /// </summary>
+        private readonly Dictionary<int, Queue<IKafkaMessage<TKey, TValue>>> _messagesNotYetCommittedByPartition;
+        private readonly ReaderWriterLockSlim _messagesNotYetCommittedByPartitionReaderWriterLock;
+
+        private readonly SemaphoreSlim _canQueueMessage;
+        private readonly CancellationToken _stopToken;
+
+        public CommitState(int maxMessagesQueued, CancellationToken stopToken)
+        {
+            this._canQueueMessage = new(maxMessagesQueued);
+            this._stopToken = stopToken;
+            this._messagesNotYetCommittedByPartition = new();
+            this._messagesNotYetCommittedByPartitionReaderWriterLock = new();
+        }
+
+        /// <summary>
+        /// Returns an enumeration of messages that are ready to be committed to kafka
+        /// </summary>
+        public IEnumerable<IKafkaMessage<TKey, TValue>> GetMessagesToCommit()
+        {
+            List<Queue<IKafkaMessage<TKey, TValue>>> allQueues;
+
+            this._messagesNotYetCommittedByPartitionReaderWriterLock.EnterReadLock();
+            try
+            {
+                allQueues = this._messagesNotYetCommittedByPartition.Values.ToList();
+            }
+            finally
+            {
+                this._messagesNotYetCommittedByPartitionReaderWriterLock.ExitReadLock();
+            }
+
+            foreach (var queue in allQueues)
+            {
+                IKafkaMessage<TKey, TValue> messageToCommit = null;
+
+                lock (queue)
+                {
+                    while (queue.TryPeek(out var msg) && msg.WasHandled)
+                    {
+                        if (messageToCommit == null || msg.Offset.Offset > messageToCommit.Offset.Offset)
```

Comment:
This partition queue should be ordered by offset ascending. We could change the `>` check to an assertion w/ logging and just always set `messageToCommit = msg;`

Reply:
If we have the assertion, we don't save any CPU cycles, might as well leave it as is.
But, I'd prefer to just remove the offset check.